op-chain-ops/cmd/celo-migrate/main.go

```diff
@@ -0,0 +1,456 @@
+		dbPathFlag,
+		dbCacheFlag,
+		dbHandlesFlag,
+		dryRunFlag,
+	}
+
+	// from `packages/contracts-bedrock/deploy-config/internal-devnet.json`
+	EIP1559Denominator = uint64(50) // TODO(pl): select values
+	EIP1559Elasticity  = uint64(10)
+)
+
+func main() {
+	log.Root().SetHandler(log.StreamHandler(os.Stderr, log.TerminalFormat(isatty.IsTerminal(os.Stderr.Fd()))))
+
+	app := &cli.App{
+		Name:  "migrate",
+		Usage: "Migrate Celo state to a CeL2 DB",
+		Flags: flags,
+		Action: func(ctx *cli.Context) error {
+			deployConfig := ctx.Path("deploy-config")
+			if deployConfig == "" {
+				return fmt.Errorf("must specify --deploy-config")
+			}
+			log.Info("Deploy config", "path", deployConfig)
+			config, err := genesis.NewDeployConfig(deployConfig)
+			if err != nil {
+				return err
+			}
+
+			// Try reading the L1 deployment information
+			l1Deployments := ctx.Path("l1-deployments")
+			if l1Deployments == "" {
+				return fmt.Errorf("must specify --l1-deployments")
+			}
+			deployments, err := genesis.NewL1Deployments(l1Deployments)
+			if err != nil {
+				return fmt.Errorf("cannot read L1 deployments at %s: %w", l1Deployments, err)
+			}
+			config.SetDeployments(deployments)
+
+			// Get latest block information from L1
+			l1RPC := ctx.String("l1-rpc")
+			if l1RPC == "" {
+				return fmt.Errorf("must specify --l1-rpc")
+			}
+
+			outfileL2 := ctx.Path("outfile.l2")
+			if outfileL2 == "" {
+				return fmt.Errorf("must specify --outfile.l2")
+			}
+
+			outfileRollup := ctx.Path("outfile.rollup")
+			if outfileRollup == "" {
+				return fmt.Errorf("must specify --outfile.rollup")
+			}
+
+			var l1StartBlock *types.Block
+			client, err := ethclient.Dial(l1RPC)
+			if err != nil {
+				return fmt.Errorf("cannot dial %s: %w", l1RPC, err)
+			}
+
+			if config.L1StartingBlockTag == nil {
+				l1StartBlock, err = client.BlockByNumber(context.Background(), nil)
+				if err != nil {
+					return fmt.Errorf("cannot fetch latest block: %w", err)
+				}
+				tag := rpc.BlockNumberOrHashWithHash(l1StartBlock.Hash(), true)
+				config.L1StartingBlockTag = (*genesis.MarshalableRPCBlockNumberOrHash)(&tag)
+			} else if config.L1StartingBlockTag.BlockHash != nil {
+				l1StartBlock, err = client.BlockByHash(context.Background(), *config.L1StartingBlockTag.BlockHash)
+				if err != nil {
+					return fmt.Errorf("cannot fetch block by hash: %w", err)
+				}
+			} else if config.L1StartingBlockTag.BlockNumber != nil {
+				l1StartBlock, err = client.BlockByNumber(context.Background(), big.NewInt(config.L1StartingBlockTag.BlockNumber.Int64()))
+				if err != nil {
+					return fmt.Errorf("cannot fetch block by number: %w", err)
+				}
+			}
+
+			// Ensure that there is a starting L1 block
+			if l1StartBlock == nil {
+				return fmt.Errorf("no starting L1 block")
+			}
+
+			// Sanity check the config. Do this after filling in the L1StartingBlockTag
+			// if it is not defined.
+			if err := config.Check(); err != nil {
+				return err
+			}
```

Comment:
There doesn't seem to be any reason to do this after retrieving l1StartBlock, since the only relevant check it makes is that L1StartingBlockTag is not nil. I think this would be neater (and faster fail for the user) if the config check came up front, and a function could be extracted to retrieve the l1 block.

Reply:
I'm not following, this isn't upstream code?

Reply:
It is upstream code.

And what I wanted to convey is that I'd like to keep it as close to the upstream as possible before the next rebase in order to reduce merge conflicts.

Reply:
I see that `config.Check` is code that exists in upstream, however I'm still somewhat confused about how a conflict would arise here, since this file `op-chain-ops/cmd/celo-migrate/main.go` does not exist in upstream?

Reply:
I'm not concerned about the code in the `main.go` file, but the code in the implementation of `Check`.

Reply:
Ahh ok, I wasn't suggesting changing Check, just placing it differently in this file.